### DISCUSSION TITLE
pkg_resources.DistributionNotFound: The 'workalendar' distribution was not found and is required by the application

### DISCUSSION THREAD

**Discussion by Davidon3** [pyinstaller/pyinstaller#6072] [Help] [answered]
After checking on other Issues, I couldn't find a way to solve my error.

## Description of the issue
I'm having an error executing my main.exe file:
```
Traceback (most recent call last):
  File "main.py", line 13, in <module>
  File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
  File "<frozen importlib._bootstrap>", line 986, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 680, in _load_unlocked
  File "PyInstaller\loader\pyimod03_importers.py", line 546, in exec_module
  File "workalendar\__init__.py", line 5, in <module>
  File "pkg_resources\__init__.py", line 466, in get_distribution
  File "pkg_resources\__init__.py", line 342, in get_provider
  File "pkg_resources\__init__.py", line 886, in require
  File "pkg_resources\__init__.py", line 772, in resolve
pkg_resources.DistributionNotFound: The 'workalendar' distribution was not found and is required by the application
[8816] Failed to execute script 'main' due to unhandled exception!
```

I made the executable through:

`pyinstaller main_MRP_v0-1.py -D --additional-hooks-dir=hooks`

my hooks directory contains this file:

````
from PyInstaller.utils.hooks import copy_metadata
datas = copy_metadata('workalendar')
````

I also tried with a similar one:
````
from PyInstaller.utils.hooks import collect_all
datas, binaries, hiddenimports = collect_all('workalendar')
````

But had the same result.

### Context information

* Output of `pyinstaller --version`: ```4.4```
* Version of Python: 3.7.8
* Platform: Windows
* How you installed Python: python.org/downloads

I'm still new on python, I need help in a newbie language

**Comment by rokm** [comment]
> my hooks directory contains this file:

What is "this file" called? Is it `hook-workalendar.py` or something else?

But either way, you can avoid using the hook and use `pyinstaller main_MRP_v0-1.py -D --copy-metadata workalendar` instead.

**Comment by Davidon3** [comment]
> What is "this file" called? Is it hook-workalendar.py or something else?

It's called hook-2-workalendar.py

> But either way, you can avoid using the hook and use pyinstaller main_MRP_v0-1.py -D --copy-metadata workalendar instead.

Thanks a lot! the workalendar issue is gone, now I have a brand new Error. I suppose there will be more like this.

```
ModuleNotFoundError: No module named 'pymssql._mssql'
[84864] Failed to execute script 'main_MRP_v0-1' due to unhandled exception!
```

How can I solve all of them at the same time?

**Comment by Davidon3** [comment]
Using a simpler version, I get a
`ModuleNotFoundError: No module named 'functions'`

I should import my custom modules in different ways?
When launch `python main_MRP_v0-1.py` everything works perfectly.
I have different versions of python and I don't have a pyinstaller.exe in the project folder. I added some paths, but maybe they're overlapping each other in some way?
Can I paste a pyinstaller.exe in the project folder?

**Comment by rokm** [comment, marked as answer]
> It's called hook-2-workalendar.py

Hooks need to be named `hook-modulename.py`, otherwise they don't get executed.

> Thanks a lot! the workalendar issue is gone, now I have a brand new Error. I suppose there will be more like this.
> 
> ```
> ModuleNotFoundError: No module named 'pymssql._mssql'
> [84864] Failed to execute script 'main_MRP_v0-1' due to unhandled exception!
> ```

This one should be solved by adding `pymssql._mssql` to hiddenimports, i.e., adding `--hiddenimport pymssql._mssql` to your PyInstaller command. Or, if there are more missing modules from `pymssql`, you can use `--collect-submodules pymssql` instead.


> Using a simpler version, I get a
> `ModuleNotFoundError: No module named 'functions'`

What is `functions`? If it is your custom module or a package, you need to ensure that it is visible to PyInstaller during the freezing process (i.e., either in PYTHONPATH, or explicitly passed to PyInstaller via `--paths`). But if it is placed next to your entry-point script, it should be picked up automatically, as long as your code explicitly imports it (you can try adding `--hiddenimport functions`, though).


> I have different versions of python and I don't have a pyinstaller.exe in the project folder. I added some paths, but maybe they're overlapping each other in some way?
> Can I paste a pyinstaller.exe in the project folder?

No, you need to ensure that PyInstaller is properly installed for the version of python that you are using. If you have multiple versions and/or environments installed, then figure out which one you're using and make sure everything (PyInstaller and 3rd party packages used by your code) are installed there. Running pyinstaller via `python3 -m PyInstaller` might be easier to debug than running `pyinstaller.exe`...

**Reply by Davidon3** [reply]
Thanks a lot for your answer.
I still have an issue about this:

> What is functions?

It's (it was) one of my custom modules. 
I checked if it's going to be imported correctly and I'll try to explain how it's done:

The main file imports a function in a module placed in a subdirectory. 
In this folder there is a foo.py and inside this file there is the function i'm importing. This funcion needs another module, which is in the same directory as foo.py.
In an older version, that extra module was called functions.py 
Now it has a different name, let's say bar.py.

In the foo.py there is the line:

`from folder.bar import function_i_need_in_foo`

after I create the executable using 
`python -m PyInstaller main_MRP_v0-1.py --copy-metadata workalendar --paths C:\*project-directory*`
and executing 
`C:\*project-directory*\dist\main_MRP_v0-1.exe`
I still get the error, as if it's refering to an older version of foo.py:
`ModuleNotFoundError: No module named 'functions' `

**Reply by rokm** [reply]
Can you try adding `--clean` to your PyInstaller command, just in case? If you keep getting the same error afterwards, double-check the source file and line indicated by the error message (you can also post the whole traceback here, perhaps that will help to shed some light on what's going on there).

**Reply by rokm** [reply]
> after I create the executable using
python -m PyInstaller main_MRP_v0-1.py --copy-metadata workalendar --paths C:\*project-directory*
and executing
C:\*project-directory*\dist\main_MRP_v0-1.exe

Hmmm, another thing. With your PyInstaller command, you seem to be building a `onedir` application, but the executable you are trying to run appear to be a `onefile` build (judging from the path). Can you remove the whole `dist` directory before rebuilding the application again, to ensure that you aren't trying to run an old build?

**Reply by Davidon3** [reply]
I tried with your suggestions, I deleted the build and dist folders and ran in windows terminal:

`python -m PyInstaller main_MRP_v0-1.py --copy-metadata workalendar --clean --onefile`

And then execute:

`C:\*project-directory*\dist\main_MRP_v0-1.exe`

I still have this error:

`Traceback (most recent call last):
  File "main_MRP_v0-1.py", line 20, in <module>
    from folder.foo import my_function
  File "PyInstaller\loader\pyimod03_importers.py", line 546, in exec_module
  File "folder\foo.py", line 5, in <module>
    from new_functions import support_function1, support_function2
ModuleNotFoundError: No module named 'functions'
[69060] Failed to execute script 'main_MRP_v0-1' due to unhandled exception!`

I double checked the source file and I haven't find any `functions` import or quote or anything.

When I run the python file normally:

`python main_MRP_v0-1.py connection.json configuration.json`

it works correctly.

**Reply by rokm** [reply]
Hmm, odd. Can you attach the `build\<name>\Analysis-00.toc` file here? Or check it yourself to see if the modules involved in the error are indeed collected from expected locations?

**Reply by Davidon3** [reply]
I renamed (with PyCharm refractor) some files and directories. Now the error is:

`C:\Users\davide.trevisiol\Documents\Re-Call\Call>dist\main_MRP_v0-1.exe
Traceback (most recent call last):
  File "main_MRP_v0-1.py", line 20, in <module>
    from MRP.MRP import MRP_function
  File "PyInstaller\loader\pyimod03_importers.py", line 546, in exec_module
  File "MRP\MRP.py", line 5, in <module>
    from MRP.funzioniMRP import genera_ordine_produzione_with_explode
ModuleNotFoundError: No module named 'funzioni'
[56508] Failed to execute script 'main_MRP_v0-1' due to unhandled exception!`

 Youl'll find the file attached 

[Analysis-00.zip](https://github.com/pyinstaller/pyinstaller/files/6931966/Analysis-00.zip)

**Reply by rokm** [reply]
So based on the Analysis-00.toc, your project layout is as follows:

Your entry-point script is in:
``` 
C:\\Users\\davide.trevisiol\\Documents\\Re-Call\\Call\\main_MRP_v0-1.py
``` 

There's `main_functions` module in same directory as entry-point script:
``` 
'C:\\Users\\davide.trevisiol\\Documents\\Re-Call\\Call\\main_functions.py'
```

And two additional packages, `Planner` and `Scheduler`, in the same directory as entry-point script:
``` 
'C:\\Users\\davide.trevisiol\\Documents\\Re-Call\\Call\\Planner\\__init__.py'
'C:\\Users\\davide.trevisiol\\Documents\\Re-Call\\Call\\Planner\\funzioniPlanner.py'
'C:\\Users\\davide.trevisiol\\Documents\\Re-Call\\Call\\Planner\\Planner.py'
'C:\\Users\\davide.trevisiol\\Documents\\Re-Call\\Call\\Scheduler\\scheduler_functions.py'
'C:\\Users\\davide.trevisiol\\Documents\\Re-Call\\Call\\Scheduler\\__init__.py'
```

And there's an `MRP` package one directory higher:
``` 
'C:\\Users\\davide.trevisiol\\Documents\\Re-Call\\MRP\\__init__.py'
'C:\\Users\\davide.trevisiol\\Documents\\Re-Call\\MRP\\MRP.py'
```

So you probalby had to either add `'C:\\Users\\davide.trevisiol\\Documents\\Re-Call`
to `PYTHONPATH`, or used PyInstaller's `--paths` to add it to the list.

And according to the error message, line 5 in 
```
'C:\\Users\\davide.trevisiol\\Documents\\Re-Call\\MRP\\MRP.py'
```
is trying to do
```
from MRP.funzioniMRP import genera_ordine_produzione_with_explode
```
whereas there does not seem to be a
`MRP.funzioniMRP` module ('C:\\Users\\davide.trevisiol\\Documents\\Re-Call\\MRP\\funzioniMRP.py').
At least not that PyInstaller's analysis is aware of it.

Can you double check if that's true? This in itself might be the cause of error, although I'm confused why the error message mentions just `funzioni` as a missing module...

**Reply by Davidon3** [reply]
> And there's an `MRP `package one directory higher:

The issue was exactly that directory: 
I had a previous version of the package outside the `Call` directory and it was added to `PATH` and had an `__init__.py` file. 
That version had a reference to `funzioni`, while the `MRP` module inside the `Call` folder is updated and has the new funcions and so on.

The problem was solved removing the higher `MRP` directory.

Thanks a lot for the help, now the executable is working.
**Please let me know how I can support the PyInstaller project.**
